command-expansion-server/src/getActivityTypescript.ts

```diff
@@ -0,0 +1,131 @@
+import {GraphQLClient, gql} from 'graphql-request';
+import {globalDeclaration, indent, interfaceDeclaration} from "./packages/lib/CodegenHelpers.js";
+import {ErrorWithStatusCode} from "./utils/ErrorWithStatusCode.js";
+
+enum SchemaTypes {
+  Int = 'int',
+  Real = 'real',
+  Duration = 'duration',
+  Boolean = 'boolean',
+  String = 'string',
+  Series = 'series',
+  Struct = 'struct',
+  Variant = 'variant',
+}
+
+type Schema =
+  | IntSchema
+  | RealSchema
+  | DurationSchema
+  | BooleanSchema
+  | StringSchema
+  | SeriesSchema
+  | StructSchema
+  | VariantSchema
+  ;
+
+interface BaseSchema<T extends SchemaTypes | unknown> {
+  type: T,
+}
+
+type StringSchema = BaseSchema<SchemaTypes.String>;
+type BooleanSchema = BaseSchema<SchemaTypes.Boolean>;
+type IntSchema = BaseSchema<SchemaTypes.Int>;
+type RealSchema = BaseSchema<SchemaTypes.Real>;
+type DurationSchema = BaseSchema<SchemaTypes.Duration>;
+
+interface SeriesSchema extends BaseSchema<SchemaTypes.Series> {
+  items: Schema,
+}
+
+interface StructSchema extends BaseSchema<SchemaTypes.Struct> {
+  items: {
+    [key: string]: Schema,
+  }
+}
+
+interface VariantSchema extends BaseSchema<SchemaTypes.Variant> {
+  variants: {
+    key: string,
+    label: string,
+  }[]
+}
```

Comment:
You are at Typescript God!!!!

Super minor, can you order this to be top-down declaration? It helps with readability.

```
enum SchemaTypes {
  Int = "int",
  Real = "real",
  Duration = "duration",
  Boolean = "boolean",
  String = "string",
  Series = "series",
  Struct = "struct",
  Variant = "variant",
}

interface BaseSchema<T extends SchemaTypes | unknown> {
  type: T;
}
interface SeriesSchema extends BaseSchema<SchemaTypes.Series> {
  items: Schema;
}

interface StructSchema extends BaseSchema<SchemaTypes.Struct> {
  items: {
    [key: string]: Schema;
  };
}

interface VariantSchema extends BaseSchema<SchemaTypes.Variant> {
  variants: {
    key: string;
    label: string;
  }[];
}

type StringSchema = BaseSchema<SchemaTypes.String>;
type BooleanSchema = BaseSchema<SchemaTypes.Boolean>;
type IntSchema = BaseSchema<SchemaTypes.Int>;
type RealSchema = BaseSchema<SchemaTypes.Real>;
type DurationSchema = BaseSchema<SchemaTypes.Duration>;


type Schema =
  | IntSchema
  | RealSchema
  | DurationSchema
  | BooleanSchema
  | StringSchema
  | SeriesSchema
  | StructSchema
  | VariantSchema;
```

Reply:
I think because TS doesn't care much about WHERE the declarations are, many libraries do it this way so that the root or most important types are higher up (seen sooner when looking through the file) and then the needed children come after them.

I'm good either way though, feel free to let me know which you prefer given the above.

Reply:
See https://github.com/graphql/graphql-js/blob/main/src/language/ast.ts#L139 as a good example of general typing patterns